modules/dlm/src/main/java/org/elasticsearch/dlm/ModifyDataLifecycleService.java

```diff
@@ -0,0 +1,151 @@
+     */
+    public void setLifecycle(
+        final List<String> dataStreamNames,
+        DataLifecycle lifecycle,
+        TimeValue ackTimeout,
+        final ActionListener<AcknowledgedResponse> listener
+    ) {
+        // TODO use request.masterNodeTimeout() as timeout?
+        taskQueue.submitTask("set-lifecycle", new ModifyLifecycleTask(dataStreamNames, lifecycle, ackTimeout, listener), null);
+    }
+
+    /**
+     * Submits the task to remove the lifecycle from the requested data streams.
+     */
+    public void removeLifecycle(List<String> dataStreamNames, TimeValue ackTimeout, ActionListener<AcknowledgedResponse> listener) {
+        // TODO use request.masterNodeTimeout() as timeout?
+        taskQueue.submitTask("delete-lifecycle", new ModifyLifecycleTask(dataStreamNames, null, ackTimeout, listener), null);
+    }
+
+    /**
+     * Creates an updated cluster state in which the requested data streams have the data lifecycle provided.
+     * Visible for testing.
+     */
+    ClusterState modifyLifecycle(ClusterState currentState, List<String> dataStreamNames, @Nullable DataLifecycle dataLifecycle) {
+        Metadata updatedMetadata = currentState.metadata();
+        Metadata.Builder builder = Metadata.builder(updatedMetadata);
+        for (var dataStreamName : dataStreamNames) {
+            var dataStream = validateDataStream(updatedMetadata, dataStreamName);
+            builder.put(
+                new DataStream(
+                    dataStream.getName(),
+                    dataStream.getIndices(),
+                    dataStream.getGeneration(),
+                    dataStream.getMetadata(),
+                    dataStream.isHidden(),
+                    dataStream.isReplicated(),
+                    dataStream.isSystem(),
+                    dataStream.isAllowCustomRouting(),
+                    dataStream.getIndexMode(),
+                    dataLifecycle
+                )
+            );
+        }
+        updatedMetadata = builder.build();
+        return ClusterState.builder(currentState).metadata(updatedMetadata).build();
+    }
+
+    private static DataStream validateDataStream(Metadata metadata, String dataStreamName) {
+        IndexAbstraction dataStream = metadata.getIndicesLookup().get(dataStreamName);
+        if (dataStream == null || dataStream.getType() != IndexAbstraction.Type.DATA_STREAM) {
+            throw new IllegalArgumentException("data stream [" + dataStreamName + "] not found");
+        }
+        return (DataStream) dataStream;
+    }
+
+    /**
+     * A cluster state update task that consists of the cluster state request and the listeners that need to be notified upon completion.
+     */
+    record ModifyLifecycleTask(
+        List<String> dataStreamNames,
+        @Nullable DataLifecycle dataLifecycle,
+        TimeValue ackTimeout,
+        ActionListener<AcknowledgedResponse> listener
+    ) implements ClusterStateTaskListener, ClusterStateAckListener {
```

Comment:
Could we extend `AckedClusterStateUpdateTask` instead?

Reply:
Unfortunately not, the `AckedClusterStateUpdateTask` is build for the old unbatched way of processing cluster updates. We could introduce a new one but we shouldn't use that one. Shall we add like this and in a follow up I can see which other places are using this (like aliases from the top of my mind) and then refactor this?

Reply:
I introduced a new abstraction for now and in a follow up PR, we can reuse it in other places too.